### ISSUE TITLE
Add vertical scroll when only 1 log file is being processed

### ISSUE THREAD

**Issue by kiniggit** [sevdokimov/log-viewer#40]

**Comment by kiniggit** [comment]
I believe we could do something like this https://www.youtube.com/watch?v=1EeVzn02grw . I mean, the lazy load could be active both to load content upwards and downwards.

**Comment by kiniggit** [comment]
I've started working on this one, okay? Whenever you have the time, add me as a contributor so I'm able to assign issues to myself.

**Comment by sevdokimov** [comment]
Good! I have added you to contributors several days ago, but you are in the pending state.
![Screenshot from 2021-04-12 12-31-42](https://user-images.githubusercontent.com/669696/114373302-20eb8f00-9b8b-11eb-9f46-3a8b0c95efaf.png)

**Comment by sevdokimov** [comment]
Hmmm... there is a problem with the scroll if a filter is defined. You can add scroll based on the file size and positions of the first and last visible records. (the file size is in `ViewStateService.statuses[logId].size` , List of loaded records is in `LogViewComponent.m`) But the file size is not relevant to user's view when a filter is defined. Let's imagine a user has defined a filter that hides half of the records and the first visible record is in the middle of the file. If the user scrolls to the beginning, he will see the first visible record at the top, he will not able to scroll above, but the scroll will show that he is in the middle of the file

**Comment by sevdokimov** [comment]
Done

**Comment by kiniggit** [comment]
> Hmmm... there is a problem with the scroll if a filter is defined. You can add scroll based on the file size and positions of the first and last visible records. (the file size is in `ViewStateService.statuses[logId].size` , List of loaded records is in `LogViewComponent.m`) But the file size is not relevant to user's view when a filter is defined. Let's imagine a user has defined a filter that hides half of the records and the first visible record is in the middle of the file. If the user scrolls to the beginning, he will see the first visible record at the top, he will not able to scroll above, but the scroll will show that he is in the middle of the file

Okay, I'm going to simulate this scenario and test how the scroll would behave. I think that we have two main cases here to talk about, in the first one the scroll is with respect to the whole log and in the second one the scroll is with respect to only the records that are currently loaded. 

I think that the scroll behavior that would make more sense to the user would be the second one, take a look at this https://infinite-scroll.com/demo/full-page/page2.html. Imagine that in the log file case the infinite behavior would be upwards and downwards (in case the log is still being generated). That way if you filter the file or proceed with some action that would shrink or expand the log content, the scroll would adjust itself. What do you think? 

I was also considering a case in which you don't remove the head or tail of the records as you're doing now. I mean, if a record has been loaded it should continue in memory. I think that this behavior of removing / not removing the records could be something configurable. What do you think?

**Comment by kiniggit** [comment]
> Done

thank you!

**Comment by kiniggit** [comment]
@sevdokimov could you take a look at https://github.com/kiniggit/log-viewer/tree/add_y_scroll_one_log . 

As I've started working with the idea of simply using "overflow-y: scroll" in logPane and check the consequences of it I've decided to first better understand the logic behind handling and showing the records.

I propose that we scroll the logPane (it doesn't matter if overflow is hidden) instead of shifting the view with marginTop. I believe that this way it will be easier to separate concerns regarding the view and the model behind it. Also, it makes it easier to think about possible extensions without having to worry about side effects that could affect the view shift  that was handled in ngAfterViewChecked. 

The branch is just a draft with the following actions tested: up, down, pageUp, pageDown, scrollBegin, scrollEnd. They were tested with and without filters. 

Let me know what you think and we can discuss this and then later we could discuss the scroll behavior with respect to the whole file or only what is loaded.

**Comment by sevdokimov** [comment]
>  and in the second one the scroll is with respect to only the records that are currently loaded.

From the user point of view no matter which records are loaded and which are not loaded. Records are loaded and unloaded automatically when user are navigating. When you read a big text file using "less" linux command, you don't care which lines are loaded to the memory and which are not loaded.

> I mean, if a record has been loaded it should continue in memory.

Browser's memory is not infinite. I'm afraid the browser can eat all machine memory. A log may be very huge.

**Comment by sevdokimov** [comment]
> I propose that we scroll the logPane (it doesn't matter if overflow is hidden) instead of shifting the view with marginTop.

Interesting, I didn't know that scroll could be used! 

> Another thing is that the view height/area of one client is not necessarily equal to another one, so I was testing the copy of permalink with shift view from one browser to another and I also think that it would be easier to work with the record index in that aswell.

How the using record index can help? The height of each record will be different on another machine if the screen weight is different.

**Comment by kiniggit** [comment]
> From the user point of view no matter which records are loaded and which are not loaded. Records are loaded and unloaded automatically when user are navigating. When you read a big text file using "less" linux command, you don't care which lines are loaded to the memory and which are not loaded.

Yes, I understand that from the user's point of view it shouldn't matter. I think that if you're working with a huge log in which you don't care about the start or end of the log, but you're navigating, searching for a term, then working with only a subset of records is ideal. Therefore, the scroll could be used only with respect to that subset of records. On the other hand, when using the component to load a small log file (several MB), or a series of log files, each one in a tab, for instance, the application could choose to load all log files in memory. Therefore, the scroll should take the whole log file into account.

**Comment by sevdokimov** [comment]
I'm not clear what do you mean. Navigation/search doesn't depend on which records are loaded and which not, it uses the whole log file. If a user wants to work with a certain region of the log, he can add filters by date (right-click on the first record in the region and select "Hide events before", then right-click on the last record in the region and select "Hide events after"). 

BTW have you realized that there is the context menu on right-click?

**Comment by kiniggit** [comment]
> How the using record index can help? The height of each record will be different on another machine if the screen weight is different.

We could postpone view computation and changes (shift, scroll, etc.) in the view to the last minute possible. For instance, the client could compute the desired scroll taking the loaded records' heights into account and then scroll to the desired record.

I don't know if the following could be an overkill, but it might help... In my day job, one of the apps that I work with is what could be considered an image viewer, in that we have two concepts:  "virtual index" and a "real index". The first one is with respect to the view, the second one is with respect to the model. They might, and usually are, be different. But it's easy to go from one context to the other .

**Comment by kiniggit** [comment]
> I'm not clear what do you mean. Navigation/search doesn't depend on which records are loaded and which not, it uses the whole log file. If a user wants to work with a certain region of the log, he can add filters by date (right-click on the first record in the region and select "Hide events before", then right-click on the last record in the region and select "Hide events after").

Sorry, my bad. I was trying to separate the context in which the user is working with a subset of records and the context in which the user works/sees the whole log.

> BTW have you realized that there is the context menu on right-click?

Yes, that is great!

**Comment by kiniggit** [comment]
whenever you want to skype let me know

**Comment by sevdokimov** [comment]
I'm ready now. You can call me now or at any time in 2h

**Comment by kiniggit** [comment]
I've just sent you a message in skype. Mine is wilson.leao.neto

**Comment by source-ram** [comment]
+1 any one working on this ?
